### DISCUSSION TITLE
Support generating custom iterators using yield keyword?

### DISCUSSION THREAD

**Discussion by danielbodart** [vlang/v#12881] [Ideas]
How do you feel about the `yield` keyword and the iterators that it generates?

Or do you feel the defer syntax gets you close enough to not warrant the additional complexity in the compiler?

If there answer was no thanks, is there any way to extend V without having to have access to the compile code? (Macro etc)

Having written hundreds of iterators in languages with and without yield syntax, I have to admit a strong preference in readability and maintainability in languages that support them. 

If there was any interest, I'd be happy to try my hand at a pull request / fork etc.

**Comment by JalonSolov** [comment]
V will (hopefully) never have macros.  The guy who invented V hates them as much as I do, so that's a good bet.  We have both been burned too many times trying to debug macro messes.  For the same reason, there's no "easy" method to extend the language.  Anyone looking at V code should be able to figure it out from only the V docs, not the V docs, plus docs for extension 1 plus docs from extension 2 plus docs from extension N ...

As for `yield`... do you have a _solid_ use case, and proof that adding it to the language would be a worthwhile investment?  V already supports iterators, and you can create your own just by adding a `.next()` method to your own custom structs, etc.

To me, `yield` appears to just be syntactic sugar for this.

**Comment by danielbodart** [comment]
Thanks for your fast response. Totally get your stance on macros and having the language be consistent etc.

`yield` is more than just syntactic code, it's about removing boiler plate code that attracts stupid bugs.

Lets take the example in the V docs:

```
struct SquareIterator {
	arr []int
mut:
	idx int
}

fn (mut iter SquareIterator) next() ?int {
	if iter.idx >= iter.arr.len {
		return error('')
	}
	defer {
		iter.idx++
	}
	return iter.arr[iter.idx] * iter.arr[iter.idx]
}

nums := [1, 2, 3, 4, 5]
iter := SquareIterator{
	arr: nums
}
for squared in iter {
	println(squared)
}
```
This code could be replaced with something like this:

```
fn squareIterator(arr []int)  Iterator<int> {
   for a in arr {
     yield a * a   // or maybe C# style yield return a * a
   }
}

nums := [1, 2, 3, 4, 5]
iter := squareIterator(nums)

for squared in iter {
	println(squared)
}
```

As you can see the intent of the code is much easier to read, it's a lot more concise, no need to do out of bounds checks or defers. Now this is a super simple case but as you can see it allows you to write lazy code like writing eager code.
It's super useful when writing infinite generators, streaming or network code.
Understand if this is a too high level a feature for a low level systems language but I thought no harm is asking! 

Maybe this is impossible in this kind of language as what is memory layout of the generated iterator? (It took Rust quite a while before they supported yield and generators)

**Reply by jdonnerstag** [reply]
I agree. I'm a big fan of generators and yield as well. It makes creating complex iterators much easier. It can all be done with structs and next(), as well, but it's more boilerplate code.
